pkg/comp-functions/functions/vshnnextcloud/collabora.go

```diff
@@ -0,0 +1,732 @@
+	cm := &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "collabora-code-coolwsd-config",
+			Namespace: comp.GetInstanceNamespace(),
+		},
+		Data: map[string]string{
+			"coolwsd.xml": coolwsd,
+			"coolkit.xml": coolkit,
+		},
+	}
+
+	return svc.SetDesiredKubeObject(cm, comp.GetName()+"-collabora-code-coolwsd-config", runtime.KubeOptionAddLabels(labelMap))
+}
+
+func createSecretWithRSAKeys(comp *vshnv1.VSHNNextcloud, svc *runtime.ServiceRuntime) error {
+	objName := comp.GetName() + "-collabora-code-ssh-key"
+
+	currentSecret := &corev1.Secret{}
+
+	err := svc.GetObservedKubeObject(currentSecret, objName)
+	if err == nil {
+		/*
+		 this part might look weird, but it's necessary because:
+		 - we don't want to overwrite the secret if it already exists
+		 - RSA generation is really slow
+		*/
+		return svc.SetDesiredKubeObject(currentSecret, objName, runtime.KubeOptionAddLabels(labelMap))
+	}
+
+	privkey, pubkey, err := generateSSHRSAKey()
+	if err != nil {
+		return err
+	}
+
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      objName,
+			Namespace: comp.GetInstanceNamespace(),
+		},
+		Data: map[string][]byte{
+			"id_rsa":     []byte(privkey),
+			"id_rsa.pub": []byte(pubkey),
+		},
+	}
+
+	return svc.SetDesiredKubeObject(secret, objName, runtime.KubeOptionAddLabels(labelMap))
+}
+
+func createServiceAccount(comp *vshnv1.VSHNNextcloud, svc *runtime.ServiceRuntime) error {
+	sa := &corev1.ServiceAccount{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      comp.GetName() + "-collabora-code-sa",
+			Namespace: comp.GetInstanceNamespace(),
+			Labels:    map[string]string{"app": comp.GetName() + "-collabora-code"},
+		},
+	}
+	return svc.SetDesiredKubeObject(sa, comp.GetName()+"-collabora-code-sa")
+	// return SetDesiredKubeObject.SetDesiredKubeObserveObjectWithLabels(sa, comp.GetName()+"-collabora-code-sa", runtime.KubeOptionAddLabels(labelMap))
+}
+
+func createSecurityContextRole(comp *vshnv1.VSHNNextcloud, svc *runtime.ServiceRuntime) error {
+	role := &rbacv1.Role{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      comp.GetName() + "-collabora-code-role",
+			Namespace: comp.GetInstanceNamespace(),
+			Labels:    map[string]string{"app": comp.GetName() + "-collabora-code"},
+		},
+		Rules: []rbacv1.PolicyRule{
+			{
+				// pods
+				APIGroups: []string{""},
+				Resources: []string{"pods", "pods/log", "pods/status", "pods/attach", "pods/exec"},
+				Verbs:     []string{"*"},
+			},
+			{
+				// deployments
+				APIGroups: []string{"apps"},
+				Resources: []string{"deployments", "statefulsets"},
+				Verbs:     []string{"get", "delete", "watch", "list", "patch", "update", "create"},
+			},
+		},
+	}
+
+	if svc.Config.Data["isOpenshift"] == "true" {
+		role.Rules = append(role.Rules, rbacv1.PolicyRule{
+			APIGroups:     []string{"security.openshift.io"},
+			Resources:     []string{"securitycontextconstraints"},
+			ResourceNames: []string{"appcat-collabora"},
+			Verbs:         []string{"use"},
```

Comment:
I think "use" is invalid.

Reply:
https://docs.openshift.com/dedicated/authentication/managing-security-context-constraints.html#role-based-access-to-ssc_configuring-internal-oauth

"use" is a valid verb in this context

Reply:
Interesting, in the main docs where Role is described there are no such verbs...